### PR TITLE
[SPARK-38346][MLLIB] Add cache in MLlib BinaryClassificationMetrics

### PR DESCRIPTION
**Pull request by waruto210**
### What changes were proposed in this pull request?

Cache two RDDs in BinaryClassificationMetrics.scala.

### Why are the changes needed?

Two RDDs in BinaryClassificationMetrics.scala were used by diffrenet jobs but not cached when we run some example code as follow(view full description on jira):
```scala
  val pipeline: Pipeline = new Pipeline().setStages(Array(assembler, classifier))

  val Array(tr, te) = originalData.randomSplit(Array(0.7, .03), 666)
  val model = pipeline.fit(tr)
  val modelDF = model.transform(te)
  val evaluator = new BinaryClassificationEvaluator().setLabelCol(labelCol).setRawPredictionCol("prediction")
  println(evaluator.evaluate(modelDF))
```

### Does this PR introduce _any_ user-facing change?

NO
### How was this patch tested?

origin tests

### CODE REVIEW

**Comment by waruto210** [comment]
cc @zhengruifeng @imatiach-msft

**Comment by srowen** [comment]
I'm not sure this is worth it. For one, this cached RDD leaks, and is not unpersisted. It doesn't check if it's already cached. And I don't think it probably helps much as you pay the caching overhead just to avoid 1 recomputation for count(), which may be not a big deal perf-wise.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala [hunk 1]
```diff
@@ -113,7 +113,7 @@ class BinaryClassificationMetrics @Since("3.0.0") (
       } else {
         iter
       }
-    }
+    }.cache()
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:116]: you can cache the returned rdd on your own purpose. 
We should not cache it here

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala [hunk 2]
```diff
@@ -186,7 +186,7 @@ class BinaryClassificationMetrics @Since("3.0.0") (
       mergeValue = (c: BinaryLabelCounter, labelAndWeight: (Double, Double)) =>
         c += (labelAndWeight._1, labelAndWeight._2),
       mergeCombiners = (c1: BinaryLabelCounter, c2: BinaryLabelCounter) => c1 += c2
-    ).sortByKey(ascending = false)
+    ).sortByKey(ascending = false).cache()
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala:189]: this RDD `counts` seems used twice at most in this method. I think it not worthwhile to cache it.

**Comment by waruto210** [comment]
The user-level cache requires developers to understand the program characteristics themselves, and the focus here is on the cache of libraries-level (MLlib), like SPARK-16697, SPARK-16880 and SPARK-18356 in MLLib.
@zhengruifeng

**Comment by zhengruifeng** [comment]
@waruto210  normally, ml algorithm will use the cached dataset `maxIters` (by default 20~100) times, so we must cache it. In this case, I think it not worthwhile.

**Pull request closed without merging**